File: machine/ts/machine/firmware/civil_time.ts
export type BmsxCivilTime = {
	year: number;
	month: number;
	day: number;
	hour: number;
	min: number;
	sec: number;
	wday: number;
	yday: number;
	isdst: false;
};

const SECONDS_PER_DAY = 86400;
const WEEKDAYS_SHORT = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'] as const;
const WEEKDAYS_LONG = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] as const;
const MONTHS_SHORT = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'] as const;
const MONTHS_LONG = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'] as const;
const LUA_CIVIL_INT_MIN = -0x80000000;
const LUA_CIVIL_INT_MAX = 0x7fffffff;
const LUA_TIME_SAFE_INT_MIN = -0x1fffffffffffff;
const LUA_TIME_SAFE_INT_MAX = 0x1fffffffffffff;

function positiveModulo(value: number, divisor: number): number {
	const mod = value % divisor;
	return mod < 0 ? mod + divisor : mod;
}

function daysFromCivil(year: number, month: number, day: number): number {
	year -= month <= 2 ? 1 : 0;
	const era = Math.floor(year / 400);
	const yearOfEra = year - era * 400;
	const monthPrime = month + (month > 2 ? -3 : 9);
	const dayOfYear = Math.floor((153 * monthPrime + 2) / 5) + day - 1;
	const dayOfEra = yearOfEra * 365 + Math.floor(yearOfEra / 4) - Math.floor(yearOfEra / 100) + dayOfYear;
	return era * 146097 + dayOfEra - 719468;
}

function civilFromDays(days: number): { year: number; month: number; day: number } {
	days += 719468;
	const era = Math.floor(days / 146097);
	const dayOfEra = days - era * 146097;
	const yearOfEra = Math.floor((dayOfEra - Math.floor(dayOfEra / 1460) + Math.floor(dayOfEra / 36524) - Math.floor(dayOfEra / 146096)) / 365);
	let year = yearOfEra + era * 400;
	const dayOfYear = dayOfEra - (365 * yearOfEra + Math.floor(yearOfEra / 4) - Math.floor(yearOfEra / 100));
	const monthPrime = Math.floor((5 * dayOfYear + 2) / 153);
	const day = dayOfYear - Math.floor((153 * monthPrime + 2) / 5) + 1;
	const month = monthPrime + (monthPrime < 10 ? 3 : -9);
	year += month <= 2 ? 1 : 0;
	return { year, month, day };
}

export function bmsxTimestampFromLuaCivilTime(year: number, month: number, day: number, hour: number, min: number, sec: number): number {
	const monthIndex = month - 1;
	const yearOffset = Math.floor(monthIndex / 12);
	const luaYear = year + yearOffset;
	const luaMonth = monthIndex - yearOffset * 12 + 1;
	return requireLuaTimeValue(daysFromCivil(luaYear, luaMonth, day) * SECONDS_PER_DAY + hour * 3600 + min * 60 + sec);
}

function requireLuaCivilInteger(value: unknown, field: string, delta: number): number {
	const number = value as number;
	if (number !== number + 0 || number - number !== 0 || number % 1 !== 0) {
		throw new Error(`field '${field}' is not an integer`);
	}
	if (!(number >= 0 ? number - delta <= LUA_CIVIL_INT_MAX : LUA_CIVIL_INT_MIN + delta <= number)) {
		throw new Error(`field '${field}' is out-of-bound`);
	}
	return number;
}

export function requireLuaCivilTimeField(value: unknown, field: string, defaultValue: number, delta: number): number {
	if (value === null) {
		if (defaultValue < 0) {
			throw new Error(`field '${field}' missing in date table`);
		}
		return defaultValue;
	}
	return requireLuaCivilInteger(value, field, delta);
}

export function requireLuaTimeValue(value: unknown): number {
	const number = value as number;
	if (number !== number + 0 || number - number !== 0 || number % 1 !== 0) {
		throw new Error('time is not an integer');
	}
	if (number < LUA_TIME_SAFE_INT_MIN || number > LUA_TIME_SAFE_INT_MAX) {
		throw new Error('time is out-of-bound');
	}
	return number;
}

export function bmsxCivilTimeFromTimestamp(timestamp: number): BmsxCivilTime {
	const days = Math.floor(timestamp / SECONDS_PER_DAY);
	const secondsOfDay = timestamp - days * SECONDS_PER_DAY;
	const civil = civilFromDays(days);
	const hour = Math.floor(secondsOfDay / 3600);
	const min = Math.floor((secondsOfDay - hour * 3600) / 60);
	const sec = secondsOfDay - hour * 3600 - min * 60;
	const yday = days - daysFromCivil(civil.year, 1, 1) + 1;
	return {
		year: civil.year,
		month: civil.month,
		day: civil.day,
		hour,
		min,
		sec,
		wday: positiveModulo(days + 4, 7) + 1,
		yday,
		isdst: false,
	};
}

function pad(value: number, size: number): string {
	let text = Math.floor(value).toString();
	while (text.length < size) {
		text = `0${text}`;
	}
	return text;
}

function padSpace(value: number, size: number): string {
	let text = Math.floor(value).toString();
	while (text.length < size) {
		text = ` ${text}`;
	}
	return text;
}

function bmsxWeekNumberSunday(time: BmsxCivilTime): number {
	return Math.floor((time.yday - 1 + 7 - (time.wday - 1)) / 7);
}

function bmsxWeekNumberMonday(time: BmsxCivilTime): number {
	return Math.floor((time.yday - 1 + 7 - ((time.wday + 5) % 7)) / 7);
}

function isoWeeksInYear(year: number): number {
	const jan1Weekday = positiveModulo(daysFromCivil(year, 1, 1) + 3, 7) + 1;
	return jan1Weekday === 4 || (jan1Weekday === 3 && daysFromCivil(year + 1, 1, 1) - daysFromCivil(year, 1, 1) === 366) ? 53 : 52;
}

function bmsxIsoWeek(time: BmsxCivilTime): { year: number; week: number } {
	const isoWeekday = time.wday === 1 ? 7 : time.wday - 1;
	let isoYear = time.year;
	let week = Math.floor((time.yday - isoWeekday + 10) / 7);
	if (week < 1) {
		isoYear -= 1;
		week = isoWeeksInYear(isoYear);
	} else {
		const weeksInYear = isoWeeksInYear(isoYear);
		if (week > weeksInYear) {
			isoYear += 1;
			week = 1;
		}
	}
	return { year: isoYear, week };
}

export function formatBmsxCivilTime(format: string, time: BmsxCivilTime): string {
	const weekdayIndex = time.wday - 1;
	const hour12 = time.hour % 12 === 0 ? 12 : time.hour % 12;
	const ampm = time.hour < 12 ? 'AM' : 'PM';
	let output = '';
	for (let index = 0; index < format.length; index += 1) {
		const ch = format.charAt(index);
		if (ch !== '%') {
			output += ch;
			continue;
		}
		index += 1;
		let code = format.charAt(index);
		let modifier = '';
		if (code === 'E' || code === 'O') {
			modifier = code;
			index += 1;
			code = format.charAt(index);
			const valid = modifier === 'E'
				? 'cCxXyY'.indexOf(code) >= 0
				: 'deHImMSuUVwWy'.indexOf(code) >= 0;
			if (!valid) {
				throw new Error(`invalid conversion specifier '%${modifier}${code}'`);
			}
		}
		const isoWeek = code === 'G' || code === 'g' || code === 'V' ? bmsxIsoWeek(time) : null;
		switch (code) {
			case 'C': output += pad(Math.floor(time.year / 100), 2); break;
			case 'D': output += `${pad(time.month, 2)}/${pad(time.day, 2)}/${pad(time.year % 100, 2)}`; break;
			case 'F': output += `${pad(time.year, 4)}-${pad(time.month, 2)}-${pad(time.day, 2)}`; break;
			case 'G': output += pad(isoWeek!.year, 4); break;
			case 'g': output += pad(isoWeek!.year % 100, 2); break;
			case 'Y': output += pad(time.year, 4); break;
			case 'y': output += pad(time.year % 100, 2); break;
			case 'm': output += pad(time.month, 2); break;
			case 'd': output += pad(time.day, 2); break;
			case 'e': output += padSpace(time.day, 2); break;
			case 'H': output += pad(time.hour, 2); break;
			case 'M': output += pad(time.min, 2); break;
			case 'S': output += pad(time.sec, 2); break;
			case 'I': output += pad(hour12, 2); break;
			case 'R': output += `${pad(time.hour, 2)}:${pad(time.min, 2)}`; break;
			case 'r': output += `${pad(hour12, 2)}:${pad(time.min, 2)}:${pad(time.sec, 2)} ${ampm}`; break;
			case 'T': output += `${pad(time.hour, 2)}:${pad(time.min, 2)}:${pad(time.sec, 2)}`; break;
			case 'p': output += ampm; break;
			case 'a': output += WEEKDAYS_SHORT[weekdayIndex]; break;
			case 'A': output += WEEKDAYS_LONG[weekdayIndex]; break;
			case 'h': output += MONTHS_SHORT[time.month - 1]; break;
			case 'b': output += MONTHS_SHORT[time.month - 1]; break;
			case 'B': output += MONTHS_LONG[time.month - 1]; break;
			case 'j': output += pad(time.yday, 3); break;
			case 'n': output += '\n'; break;
			case 't': output += '\t'; break;
			case 'U': output += pad(bmsxWeekNumberSunday(time), 2); break;
			case 'u': output += String(time.wday === 1 ? 7 : time.wday - 1); break;
			case 'V': output += pad(isoWeek!.week, 2); break;
			case 'w': output += String(weekdayIndex); break;
			case 'W': output += pad(bmsxWeekNumberMonday(time), 2); break;
			case 'c': output += `${WEEKDAYS_SHORT[weekdayIndex]} ${MONTHS_SHORT[time.month - 1]} ${pad(time.day, 2)} ${pad(time.hour, 2)}:${pad(time.min, 2)}:${pad(time.sec, 2)} ${pad(time.year, 4)}`; break;
			case 'x': output += `${pad(time.month, 2)}/${pad(time.day, 2)}/${pad(time.year % 100, 2)}`; break;
			case 'X': output += `${pad(time.hour, 2)}:${pad(time.min, 2)}:${pad(time.sec, 2)}`; break;
			case 'z': output += '+0000'; break;
			case 'Z': output += 'BMSX'; break;
			case '%': output += '%'; break;
			default: throw new Error(`invalid conversion specifier '%${modifier}${code}'`);
		}
	}
	return output;
}

File: machine/cpp/machine/firmware/civil_time.cpp
#include "machine/firmware/civil_time.h"

#include <array>
#include <stdexcept>

namespace bmsx {
namespace {
constexpr i64 SECONDS_PER_DAY = 86'400;
constexpr std::array<std::string_view, 7> WEEKDAYS_SHORT{"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
constexpr std::array<std::string_view, 7> WEEKDAYS_LONG{"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};
constexpr std::array<std::string_view, 12> MONTHS_SHORT{"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
constexpr std::array<std::string_view, 12> MONTHS_LONG{"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"};
constexpr i64 LUA_TIME_SAFE_INT_MIN = -9'007'199'254'740'991LL;
constexpr i64 LUA_TIME_SAFE_INT_MAX = 9'007'199'254'740'991LL;

auto floorDiv(i64 value, i64 divisor) -> i64 {
	i64 quotient = value / divisor;
	const i64 remainder = value % divisor;
	if (remainder < 0) {
		quotient -= 1;
	}
	return quotient;
}

auto positiveModulo(i64 value, i64 divisor) -> i64 {
	const i64 mod = value % divisor;
	return mod < 0 ? mod + divisor : mod;
}

auto daysFromCivil(i64 year, i64 month, i64 day) -> i64 {
	year -= month <= 2 ? 1 : 0;
	const i64 era = floorDiv(year, 400);
	const i64 yearOfEra = year - era * 400;
	const i64 monthPrime = month + (month > 2 ? -3 : 9);
	const i64 dayOfYear = (153 * monthPrime + 2) / 5 + day - 1;
	const i64 dayOfEra = yearOfEra * 365 + yearOfEra / 4 - yearOfEra / 100 + dayOfYear;
	return era * 146097 + dayOfEra - 719468;
}

struct CivilDate {
	i64 year;
	int month;
	int day;
};

auto civilFromDays(i64 days) -> CivilDate {
	days += 719468;
	const i64 era = floorDiv(days, 146097);
	const i64 dayOfEra = days - era * 146097;
	const i64 yearOfEra = (dayOfEra - dayOfEra / 1460 + dayOfEra / 36524 - dayOfEra / 146096) / 365;
	i64 year = yearOfEra + era * 400;
	const i64 dayOfYear = dayOfEra - (365 * yearOfEra + yearOfEra / 4 - yearOfEra / 100);
	const i64 monthPrime = (5 * dayOfYear + 2) / 153;
	const int day = static_cast<int>(dayOfYear - (153 * monthPrime + 2) / 5 + 1);
	const int month = static_cast<int>(monthPrime + (monthPrime < 10 ? 3 : -9));
	year += month <= 2 ? 1 : 0;
	return {.year = year, .month = month, .day = day};
}

void appendPadded(std::string& output, i64 value, int size) {
	std::string text = std::to_string(value);
	while (static_cast<int>(text.size()) < size) {
		output.push_back('0');
		size -= 1;
	}
	output += text;
}

void appendSpacePadded(std::string& output, i64 value, int size) {
	std::string text = std::to_string(value);
	while (static_cast<int>(text.size()) < size) {
		output.push_back(' ');
		size -= 1;
	}
	output += text;
}

i64 weekNumberSunday(const BmsxCivilTime& time) {
	return (static_cast<i64>(time.yearday) - 1 + 7 - (time.weekday - 1)) / 7;
}

i64 weekNumberMonday(const BmsxCivilTime& time) {
	return (static_cast<i64>(time.yearday) - 1 + 7 - ((time.weekday + 5) % 7)) / 7;
}

i64 isoWeeksInYear(i64 year) {
	const i64 jan1Weekday = positiveModulo(daysFromCivil(year, 1, 1) + 3, 7) + 1;
	return jan1Weekday == 4 || (jan1Weekday == 3 && daysFromCivil(year + 1, 1, 1) - daysFromCivil(year, 1, 1) == 366) ? 53 : 52;
}

struct BmsxIsoWeek {
	i64 year;
	i64 week;
};

BmsxIsoWeek isoWeek(const BmsxCivilTime& time) {
	const int isoWeekday = time.weekday == 1 ? 7 : time.weekday - 1;
	BmsxIsoWeek out{.year = time.year, .week = (static_cast<i64>(time.yearday) - isoWeekday + 10) / 7};
	if (out.week < 1) {
		out.year -= 1;
		out.week = isoWeeksInYear(out.year);
	} else {
		const i64 weeksInYear = isoWeeksInYear(out.year);
		if (out.week > weeksInYear) {
			out.year += 1;
			out.week = 1;
		}
	}
	return out;
}

bool validExtendedSpecifier(char modifier, char code) {
	switch (modifier) {
		case 'E':
			return code == 'c' || code == 'C' || code == 'x' || code == 'X' || code == 'y' || code == 'Y';
		case 'O':
			return code == 'd' || code == 'e' || code == 'H' || code == 'I' || code == 'm' || code == 'M'
				|| code == 'S' || code == 'u' || code == 'U' || code == 'V' || code == 'w' || code == 'W'
				|| code == 'y';
		default:
			return false;
	}
}

std::runtime_error invalidConversionSpecifier(char modifier, char code) {
	std::string specifier = "%";
	if (modifier != '\0') {
		specifier.push_back(modifier);
	}
	if (code != '\0') {
		specifier.push_back(code);
	}
	return std::runtime_error("invalid conversion specifier '" + specifier + "'");
}

} // namespace

auto bmsxTimestampFromLuaCivilTime(i64 year, i64 month, i64 day, i64 hour, i64 minute, i64 second) -> i64 {
	const i64 monthIndex = month - 1;
	const i64 yearOffset = floorDiv(monthIndex, 12);
	const i64 luaYear = year + yearOffset;
	const i64 luaMonth = monthIndex - yearOffset * 12 + 1;
	const i64 timestamp = daysFromCivil(luaYear, luaMonth, day) * SECONDS_PER_DAY + hour * 3600 + minute * 60 + second;
	if (timestamp < LUA_TIME_SAFE_INT_MIN || timestamp > LUA_TIME_SAFE_INT_MAX) {
		throw std::runtime_error("time is out-of-bound");
	}
	return timestamp;
}

auto bmsxCivilTimeFromTimestamp(i64 timestamp) -> BmsxCivilTime {
	const i64 days = floorDiv(timestamp, SECONDS_PER_DAY);
	const i64 secondsOfDay = timestamp - days * SECONDS_PER_DAY;
	const CivilDate civil = civilFromDays(days);
	const int hour = static_cast<int>(secondsOfDay / 3600);
	const int minute = static_cast<int>((secondsOfDay - static_cast<i64>(hour) * 3600) / 60);
	const int second = static_cast<int>(secondsOfDay - static_cast<i64>(hour) * 3600 - static_cast<i64>(minute) * 60);
	const int yearday = static_cast<int>(days - daysFromCivil(civil.year, 1, 1) + 1);
	return {
		.year = civil.year,
		.month = civil.month,
		.day = civil.day,
		.hour = hour,
		.minute = minute,
		.second = second,
		.weekday = static_cast<int>(positiveModulo(days + 4, 7)) + 1,
		.yearday = yearday,
		.isDst = false,
	};
}

auto formatBmsxCivilTime(std::string_view format, const BmsxCivilTime& time) -> std::string {
	const int weekdayIndex = time.weekday - 1;
	const int hour12 = time.hour % 12 == 0 ? 12 : time.hour % 12;
	std::string output;
	for (size_t index = 0; index < format.size(); index += 1) {
		const char ch = format[index];
		if (ch != '%') {
			output.push_back(ch);
			continue;
		}
		index += 1;
		char code = index < format.size() ? format[index] : '\0';
		char modifier = '\0';
		if (code == 'E' || code == 'O') {
			modifier = code;
			index += 1;
			code = index < format.size() ? format[index] : '\0';
			if (!validExtendedSpecifier(modifier, code)) {
				throw invalidConversionSpecifier(modifier, code);
			}
		}
		const BmsxIsoWeek iso = (code == 'G' || code == 'g' || code == 'V') ? isoWeek(time) : BmsxIsoWeek{};
		switch (code) {
			case 'C': appendPadded(output, floorDiv(time.year, 100), 2); break;
			case 'D':
				appendPadded(output, time.month, 2);
				output.push_back('/');
				appendPadded(output, time.day, 2);
				output.push_back('/');
				appendPadded(output, time.year % 100, 2);
				break;
			case 'F':
				appendPadded(output, time.year, 4);
				output.push_back('-');
				appendPadded(output, time.month, 2);
				output.push_back('-');
				appendPadded(output, time.day, 2);
				break;
			case 'G': appendPadded(output, iso.year, 4); break;
			case 'g': appendPadded(output, iso.year % 100, 2); break;
			case 'Y': appendPadded(output, time.year, 4); break;
			case 'y': appendPadded(output, time.year % 100, 2); break;
			case 'm': appendPadded(output, time.month, 2); break;
			case 'd': appendPadded(output, time.day, 2); break;
			case 'e': appendSpacePadded(output, time.day, 2); break;
			case 'H': appendPadded(output, time.hour, 2); break;
			case 'M': appendPadded(output, time.minute, 2); break;
			case 'S': appendPadded(output, time.second, 2); break;
			case 'I': appendPadded(output, hour12, 2); break;
			case 'R':
				appendPadded(output, time.hour, 2);
				output.push_back(':');
				appendPadded(output, time.minute, 2);
				break;
			case 'r':
				appendPadded(output, hour12, 2);
				output.push_back(':');
				appendPadded(output, time.minute, 2);
				output.push_back(':');
				appendPadded(output, time.second, 2);
				output.push_back(' ');
				output.append(time.hour < 12 ? "AM" : "PM");
				break;
			case 'T':
				appendPadded(output, time.hour, 2);
				output.push_back(':');
				appendPadded(output, time.minute, 2);
				output.push_back(':');
				appendPadded(output, time.second, 2);
				break;
			case 'p': output.append(time.hour < 12 ? "AM" : "PM"); break;
			case 'a': output.append(WEEKDAYS_SHORT[weekdayIndex]); break;
			case 'A': output.append(WEEKDAYS_LONG[weekdayIndex]); break;
			case 'h': output.append(MONTHS_SHORT[time.month - 1]); break;
			case 'b': output.append(MONTHS_SHORT[time.month - 1]); break;
			case 'B': output.append(MONTHS_LONG[time.month - 1]); break;
			case 'j': appendPadded(output, time.yearday, 3); break;
			case 'n': output.push_back('\n'); break;
			case 't': output.push_back('\t'); break;
			case 'U': appendPadded(output, weekNumberSunday(time), 2); break;
			case 'u': output += std::to_string(time.weekday == 1 ? 7 : time.weekday - 1); break;
			case 'V': appendPadded(output, iso.week, 2); break;
			case 'w': output += std::to_string(weekdayIndex); break;
			case 'W': appendPadded(output, weekNumberMonday(time), 2); break;
			case 'c':
				output.append(WEEKDAYS_SHORT[weekdayIndex]);
				output.push_back(' ');
				output.append(MONTHS_SHORT[time.month - 1]);
				output.push_back(' ');
				appendPadded(output, time.day, 2);
				output.push_back(' ');
				appendPadded(output, time.hour, 2);
				output.push_back(':');
				appendPadded(output, time.minute, 2);
				output.push_back(':');
				appendPadded(output, time.second, 2);
				output.push_back(' ');
				appendPadded(output, time.year, 4);
				break;
			case 'x':
				appendPadded(output, time.month, 2);
				output.push_back('/');
				appendPadded(output, time.day, 2);
				output.push_back('/');
				appendPadded(output, time.year % 100, 2);
				break;
			case 'X':
				appendPadded(output, time.hour, 2);
				output.push_back(':');
				appendPadded(output, time.minute, 2);
				output.push_back(':');
				appendPadded(output, time.second, 2);
				break;
			case 'z': output.append("+0000"); break;
			case 'Z': output.append("BMSX"); break;
			case '%': output.push_back('%'); break;
			default:
				throw invalidConversionSpecifier(modifier, code);
		}
	}
	return output;
}

} // namespace bmsx
